File: 2d/More_complex_geometries/__epyccel__/__pyccel__/mod_v80bvlt5fqyi.f90
module mod_v80bvlt5fqyi


  use, intrinsic :: ISO_C_Binding, only : f64 => C_DOUBLE , i64 => &
        C_INT64_T
  implicit none

  contains

  !........................................
  subroutine assemble_matrix_ex01(ne1, ne2, p1, p2, spans_1, spans_2, &
        basis_1, basis_2, weights_1, weights_2, points_1, points_2, &
        matrix)

    implicit none

    integer(i64), value :: ne1
    integer(i64), value :: ne2
    integer(i64), value :: p1
    integer(i64), value :: p2
    integer(i64), intent(in) :: spans_1(0:)
    integer(i64), intent(in) :: spans_2(0:)
    real(f64), intent(in) :: basis_1(0:,0:,0:,0:)
    real(f64), intent(in) :: basis_2(0:,0:,0:,0:)
    real(f64), intent(in) :: weights_1(0:,0:)
    real(f64), intent(in) :: weights_2(0:,0:)
    real(f64), intent(in) :: points_1(0:,0:)
    real(f64), intent(in) :: points_2(0:,0:)
    real(f64), intent(inout) :: matrix(0:,0:,0:,0:)
    integer(i64) :: k1
    integer(i64) :: k2
    real(f64), allocatable :: lcoeffs_u(:,:)
    real(f64), allocatable :: arr_J_mat0(:,:)
    real(f64), allocatable :: arr_J_mat1(:,:)
    real(f64), allocatable :: arr_J_mat2(:,:)
    real(f64), allocatable :: arr_J_mat3(:,:)
    real(f64), allocatable :: J_mat(:,:)
    integer(i64) :: ie1
    integer(i64) :: i_span_1
    integer(i64) :: ie2
    integer(i64) :: i_span_2
    integer(i64) :: g1
    integer(i64) :: g2
    real(f64) :: x1
    real(f64) :: x2
    real(f64) :: F1x
    real(f64) :: F1y
    real(f64) :: F2y
    real(f64) :: F2x
    real(f64) :: det_Hess
    integer(i64) :: il_1
    integer(i64) :: il_2
    integer(i64) :: jl_1
    integer(i64) :: jl_2
    integer(i64) :: i1
    integer(i64) :: j1
    integer(i64) :: i2
    integer(i64) :: j2
    real(f64) :: v
    real(f64) :: bi_x1
    real(f64) :: bi_x2
    real(f64) :: bj_x1
    real(f64) :: bj_x2
    real(f64) :: bi_x
    real(f64) :: bi_y
    real(f64) :: bj_x
    real(f64) :: bj_y
    real(f64) :: wvol

    k1 = size(weights_1, 1_i64, i64)
    k2 = size(weights_2, 1_i64, i64)
    !.. circle
    !...
    allocate(lcoeffs_u(0:p2 + 1_i64 - 1_i64, 0:p1 + 1_i64 - 1_i64))
    lcoeffs_u = 0.0_f64
    allocate(arr_J_mat0(0:k2 - 1_i64, 0:k1 - 1_i64))
    arr_J_mat0 = 0.0_f64
    allocate(arr_J_mat1(0:k2 - 1_i64, 0:k1 - 1_i64))
    arr_J_mat1 = 0.0_f64
    allocate(arr_J_mat2(0:k2 - 1_i64, 0:k1 - 1_i64))
    arr_J_mat2 = 0.0_f64
    allocate(arr_J_mat3(0:k2 - 1_i64, 0:k1 - 1_i64))
    arr_J_mat3 = 0.0_f64
    allocate(J_mat(0:k2 - 1_i64, 0:k1 - 1_i64))
    J_mat = 0.0_f64
    !... build matrices
    do ie1 = 0_i64, ne1 - 1_i64
      i_span_1 = spans_1(ie1)
      do ie2 = 0_i64, ne2 - 1_i64
        i_span_2 = spans_2(ie2)
        do g1 = 0_i64, k1 - 1_i64
          do g2 = 0_i64, k2 - 1_i64
            x1 = points_1(g1, ie1)
            x2 = points_2(g2, ie2)
            F1x = 2.0_f64 * sqrt(1.0_f64 - 0.5_f64 * (2.0_f64 * x2 - &
                  1.0_f64) ** 2_i64)
            F1y = (-(2.0_f64 * x1 - 1.0_f64)) * (2.0_f64 * x2 - 1.0_f64 &
                  ) / sqrt(1.0_f64 - 0.5_f64 * (2.0_f64 * x2 - 1.0_f64 &
                  ) ** 2_i64)
            F2y = 2.0_f64 * sqrt(1.0_f64 - 0.5_f64 * (2.0_f64 * x1 - &
                  1.0_f64) ** 2_i64)
            F2x = (-(2.0_f64 * x1 - 1.0_f64)) * (2.0_f64 * x2 - 1.0_f64 &
                  ) / sqrt(1.0_f64 - 0.5_f64 * (2.0_f64 * x1 - 1.0_f64 &
                  ) ** 2_i64)
            det_Hess = abs(F1x * F2y - F1y * F2x)
            !F=(F1, F2)
            arr_J_mat0(g2, g1) = F2y
            arr_J_mat1(g2, g1) = F1x
            arr_J_mat2(g2, g1) = F1y
            arr_J_mat3(g2, g1) = F2x
            J_mat(g2, g1) = det_Hess
          end do
        end do
        do il_1 = 0_i64, p1 + 1_i64 - 1_i64
          do il_2 = 0_i64, p2 + 1_i64 - 1_i64
            do jl_1 = 0_i64, p1 + 1_i64 - 1_i64
              do jl_2 = 0_i64, p2 + 1_i64 - 1_i64
                i1 = i_span_1 - p1 + il_1
                j1 = i_span_1 - p1 + jl_1
                i2 = i_span_2 - p2 + il_2
                j2 = i_span_2 - p2 + jl_2
                v = 0.0_f64
                do g1 = 0_i64, k1 - 1_i64
                  do g2 = 0_i64, k2 - 1_i64
                    bi_x1 = basis_1(g1, 1_i64, il_1, ie1) * basis_2(g2, &
                          0_i64, il_2, ie2)
                    bi_x2 = basis_1(g1, 0_i64, il_1, ie1) * basis_2(g2, &
                          1_i64, il_2, ie2)
                    bj_x1 = basis_1(g1, 1_i64, jl_1, ie1) * basis_2(g2, &
                          0_i64, jl_2, ie2)
                    bj_x2 = basis_1(g1, 0_i64, jl_1, ie1) * basis_2(g2, &
                          1_i64, jl_2, ie2)
                    bi_x = arr_J_mat0(g2, g1) * bi_x1 - arr_J_mat3(g2, &
                          g1) * bi_x2
                    bi_y = arr_J_mat1(g2, g1) * bi_x2 - arr_J_mat2(g2, &
                          g1) * bi_x1
                    bj_x = arr_J_mat0(g2, g1) * bj_x1 - arr_J_mat3(g2, &
                          g1) * bj_x2
                    bj_y = arr_J_mat1(g2, g1) * bj_x2 - arr_J_mat2(g2, &
                          g1) * bj_x1
                    wvol = weights_1(g1, ie1) * weights_2(g2, ie2)
                    v = v + (bi_x * bj_x + bi_y * bj_y) * wvol / J_mat( &
                          g2, g1)
                  end do
                end do
                matrix(p2 + j2 - i2, p1 + j1 - i1, p2 + i2, p1 + i1) = &
                      matrix(p2 + j2 - i2, p1 + j1 - i1, p2 + i2, p1 + &
                      i1) + v
              end do
            end do
          end do
        end do
      end do
    end do
    if (allocated(arr_J_mat3)) then
      deallocate(arr_J_mat3)
    end if
    if (allocated(J_mat)) then
      deallocate(J_mat)
    end if
    if (allocated(arr_J_mat2)) then
      deallocate(arr_J_mat2)
    end if
    if (allocated(arr_J_mat1)) then
      deallocate(arr_J_mat1)
    end if
    if (allocated(arr_J_mat0)) then
      deallocate(arr_J_mat0)
    end if
    if (allocated(lcoeffs_u)) then
      deallocate(lcoeffs_u)
    end if

  end subroutine assemble_matrix_ex01
  !........................................

end module mod_v80bvlt5fqyi

File: 2d/More_complex_geometries/__epyccel__/mod_v80bvlt5fqyi.py
@types('int', 'int', 'int', 'int', 'int[:]', 'int[:]', 'double[:,:,:,:]', 'double[:,:,:,:]', 'double[:,:]', 'double[:,:]', 'double[:,:]', 'double[:,:]', 'double[:,:,:,:]')
def assemble_matrix_ex01(ne1, ne2, p1, p2, spans_1, spans_2, basis_1, basis_2, weights_1, weights_2, points_1, points_2, matrix):
	from numpy import exp
	from numpy import cos
	from numpy import sin
	from numpy import pi
	from numpy import sqrt
	from numpy import zeros

	k1 = weights_1.shape[1]
	k2 = weights_2.shape[1]

	#.. circle

	# ...
	lcoeffs_u  = zeros((p1+1,p2+1))
	arr_J_mat0 = zeros((k1,k2))
	arr_J_mat1 = zeros((k1,k2))
	arr_J_mat2 = zeros((k1,k2))
	arr_J_mat3 = zeros((k1,k2))
	J_mat      = zeros((k1,k2))

    # ... build matrices
	for ie1 in range(0, ne1):
		i_span_1 = spans_1[ie1]
		for ie2 in range(0, ne2):
			i_span_2 = spans_2[ie2]

			for g1 in range(0, k1):
				for g2 in range(0, k2):

					x1       = points_1[ie1,g1]
					x2       = points_2[ie2,g2]


					F1x      = 2.0*sqrt(1.0-0.5*(2.0*x2-1.0)**2)
					F1y      = -(2.0*x1-1.0)*(2.0*x2-1.0)/sqrt(1.0-0.5*(2.0*x2-1.0)**2)
					F2y      = 2.0*sqrt(1.0-0.5*(2.0*x1-1.0)**2)
					F2x      = -(2.0*x1-1.0)*(2.0*x2-1.0)/sqrt(1.0-0.5*(2.0*x1-1.0)**2)
					det_Hess = abs(F1x*F2y-F1y*F2x)

					#F=(F1, F2)
					arr_J_mat0[g1,g2] = F2y
					arr_J_mat1[g1,g2] = F1x
					arr_J_mat2[g1,g2] = F1y
					arr_J_mat3[g1,g2] = F2x
			
					J_mat[g1,g2]      = det_Hess
		                    
			for il_1 in range(0, p1+1):
				for il_2 in range(0, p2+1):
					for jl_1 in range(0, p1+1):
				   		for jl_2 in range(0, p2+1):
				   			i1 = i_span_1 - p1 + il_1
				   			j1 = i_span_1 - p1 + jl_1
				   			i2 = i_span_2 - p2 + il_2
				   			j2 = i_span_2 - p2 + jl_2
				   			v  = 0.0
				   			for g1 in range(0, k1):
				   				for g2 in range(0, k2):
				   					
				   					bi_x1 = basis_1[ie1, il_1, 1, g1] * basis_2[ie2, il_2, 0, g2]
				   					bi_x2 = basis_1[ie1, il_1, 0, g1] * basis_2[ie2, il_2, 1, g2]
				   					
				   					bj_x1 = basis_1[ie1, jl_1, 1, g1] * basis_2[ie2, jl_2, 0, g2]
				   					bj_x2 = basis_1[ie1, jl_1, 0, g1] * basis_2[ie2, jl_2, 1, g2]
				   					
				   					bi_x = arr_J_mat0[g1,g2] * bi_x1 - arr_J_mat3[g1,g2] * bi_x2
				   					bi_y = arr_J_mat1[g1,g2] * bi_x2 - arr_J_mat2[g1,g2] * bi_x1
				   					
				   					bj_x = arr_J_mat0[g1,g2] * bj_x1 - arr_J_mat3[g1,g2] * bj_x2
				   					bj_y = arr_J_mat1[g1,g2] * bj_x2 - arr_J_mat2[g1,g2] * bj_x1
				   					
				   					wvol = weights_1[ie1, g1] * weights_2[ie2, g2]
				   					
				   					v   += (bi_x * bj_x + bi_y * bj_y ) * wvol / J_mat[g1,g2]
				   			matrix[p1+i1, p2+i2, p1+j1-i1, p2+j2-i2]  += v
